merkle/tree_hasher.go

```diff
@@ -14,11 +14,24 @@
 
 package merkle
 
-// TreeHasher is the interface that the previous tree hasher struct implemented.
+// TreeHasher computes generic merkle trees.
 type TreeHasher interface {
-	HashEmpty() []byte
+	// EmptyRoot returns a special case for the root of an empty tree.
+	EmptyRoot() []byte
+	// HashEmpty returns the hash of an empty branch at a given depth.
+	// A height of 0 indicates an empty leaf. The maximum height is Size*8.
+	HashEmpty(height int) []byte
+	// HashLeaf computes the hash of a leaf that exists.
 	HashLeaf(leaf []byte) []byte
+	// HashChildren computes interior nodes.
 	HashChildren(l, r []byte) []byte
+	// Size is the number of bits in the underlying hash function.
+	// It is also the maximum height of the merkle tree.
 	// TODO(gbelvin): Replace Size() with BitLength().
 	Size() int
 }
+
+// MapHasher computes sparse merkle trees of a static height.
```

Comment:
Should mention hashing rather than just "computes". Is this interface going to remain after all the refactoring? If not add a TODO.